packages/core/src/agent/MessageSender.ts

```diff
@@ -408,13 +419,61 @@ export class MessageSender {
     const outboundPackage = await this.packMessage(agentContext, { message, keys, endpoint: service.serviceEndpoint })
     outboundPackage.endpoint = service.serviceEndpoint
     outboundPackage.connectionId = connection?.id
-    for (const transport of this.outboundTransports) {
-      const protocolScheme = getProtocolScheme(service.serviceEndpoint)
-      if (!protocolScheme) {
-        this.logger.warn('Service does not have valid protocolScheme.')
-      } else if (transport.supportedSchemes.includes(protocolScheme)) {
-        await transport.sendMessage(outboundPackage)
-        return
+
+    // Outbound transport is present
+    if (this.outboundTransports.length !== 0) {
+      for (const transport of this.outboundTransports) {
+        const protocolScheme = getProtocolScheme(service.serviceEndpoint)
+        if (!protocolScheme) {
+          this.logger.warn('Service does not have valid protocolScheme.')
+        } else if (transport.supportedSchemes.includes(protocolScheme)) {
+          await transport.sendMessage(outboundPackage)
+          return
+        }
+      }
+      // No outbound transport: Try retrieving session-based connection instead
+    } else if (sessionId || sessionIdFromInbound) {
+      const session = sessionIdFromInbound
+        ? this.transportService.findSessionById(sessionIdFromInbound)
+        : sessionId
+        ? this.transportService.findSessionById(sessionId)
+        : undefined
+
+      if (session?.inboundMessage?.hasReturnRouting(message.threadId)) {
+        this.logger.debug(
+          `Found session with return routing for message '${message.id}' (connection '${connection?.id}'`
+        )
+        try {
+          await this.sendMessageToService(outboundMessageContext)
+          this.emitMessageSentEvent(outboundMessageContext, OutboundMessageSendStatus.SentToSession)
+          return
+        } catch (error) {
+          this.logger.debug(`Sending an outbound message via session failed with error: ${error.message}.`, error)
+          throw new MessageSendingError(
+            `Unable to send message to service: ${service.serviceEndpoint} ${JSON.stringify(
+              message.threadId
+            )} ${JSON.stringify(session)}`,
+            {
+              outboundMessageContext,
+            }
+          )
+        }
+      } else if (outboundMessageContext.serviceParams?.service.serviceEndpoint && session) {
+        try {
+          await this.sendMessageToSession(agentContext, session, message)
+          this.emitMessageSentEvent(outboundMessageContext, OutboundMessageSendStatus.SentToSession)
+          return
+        } catch (error) {
+          this.logger.debug(`Sending an outbound message via session failed with error: ${error.message}.`, error)
```

Comment:
I am not sure what the exact pattern is in AFJ here, but is it common that we log a `debug` message when we also throw an error? I would say either no log or an error log.

Reply:
Also, we already pass the `error` object to the log function, so why is the `error.message` also added to the `message`?